File: crates/core/src/provider.rs
//! One client for every runtime.
//!
//! Ollama, LM Studio, Jan and Msty all speak OpenAI-compatible `/v1`, so the
//! only thing that differs is the port and which models happen to be loaded.
//! Cloud BYOK is the same surface with a base URL and a key.

use anyhow::{Context, Result, bail};
use serde::Deserialize;
use serde_json::json;
use std::time::Duration;

/// Known local runtimes, in the order we probe them.
pub const LOCAL_ENDPOINTS: &[(&str, &str)] = &[
    ("Ollama", "http://localhost:11434/v1"),
    ("LM Studio", "http://localhost:1234/v1"),
    ("Jan", "http://localhost:1337/v1"),
    ("Msty", "http://localhost:10000/v1"),
];

#[derive(Debug, Clone)]
pub struct Provider {
    pub name: String,
    pub base_url: String,
    pub api_key: Option<String>,
    http: reqwest::blocking::Client,
}

#[derive(Deserialize)]
struct ModelList {
    data: Vec<ModelEntry>,
}

#[derive(Deserialize)]
struct ModelEntry {
    id: String,
}

#[derive(Deserialize)]
struct EmbeddingResponse {
    data: Vec<EmbeddingEntry>,
}

#[derive(Deserialize)]
struct EmbeddingEntry {
    embedding: Vec<f32>,
}

#[derive(Deserialize)]
struct ChatResponse {
    choices: Vec<Choice>,
}

#[derive(Deserialize)]
struct Choice {
    message: ChatMessage,
}

#[derive(Deserialize)]
struct ChatMessage {
    content: String,
}

impl Provider {
    pub fn new(name: impl Into<String>, base_url: impl Into<String>, api_key: Option<String>) -> Self {
        Provider {
            name: name.into(),
            base_url: base_url.into().trim_end_matches('/').to_string(),
            api_key,
            http: reqwest::blocking::Client::builder()
                // Local generation on a 24B model can genuinely take a while.
                .timeout(Duration::from_secs(300))
                .build()
                .expect("http client"),
        }
    }

    /// Probe the well-known ports and return whatever is actually running.
    /// This is what lets the app say "found LM Studio running Qwen3" on first
    /// launch instead of asking the user to configure an endpoint.
    pub fn discover() -> Vec<(Provider, Vec<String>)> {
        let probe = reqwest::blocking::Client::builder()
            .timeout(Duration::from_millis(400))
            .build()
            .expect("http client");

        LOCAL_ENDPOINTS
            .iter()
            .filter_map(|(name, url)| {
                let list: ModelList = probe.get(format!("{url}/models")).send().ok()?.json().ok()?;
                let models = list.data.into_iter().map(|m| m.id).collect();
                Some((Provider::new(*name, *url, None), models))
            })
            .collect()
    }

    fn post(&self, path: &str, body: serde_json::Value) -> Result<reqwest::blocking::Response> {
        let mut req = self.http.post(format!("{}{path}", self.base_url)).json(&body);
        if let Some(key) = &self.api_key {
            req = req.bearer_auth(key);
        }
        let res = req
            .send()
            .with_context(|| format!("{} unreachable at {}", self.name, self.base_url))?;

        if !res.status().is_success() {
            let status = res.status();
            let detail = res.text().unwrap_or_default();
            bail!("{} returned {status}: {}", self.name, detail.trim());
        }
        Ok(res)
    }

    pub fn embed(&self, model: &str, inputs: &[String]) -> Result<Vec<Vec<f32>>> {
        if inputs.is_empty() {
            return Ok(vec![]);
        }
        let res: EmbeddingResponse = self
            .post("/embeddings", json!({ "model": model, "input": inputs }))?
            .json()
            .context("malformed embedding response")?;

        if res.data.len() != inputs.len() {
            bail!("expected {} embeddings, got {}", inputs.len(), res.data.len());
        }
        Ok(res.data.into_iter().map(|e| e.embedding).collect())
    }

    /// `schema` enables constrained decoding. Small local models are unreliable
    /// at tool-calling but near-perfect when JSON is enforced at the sampler,
    /// which is why every extraction path in this project goes through here.
    pub fn chat(
        &self,
        model: &str,
        system: &str,
        user: &str,
        schema: Option<serde_json::Value>,
    ) -> Result<String> {
        let mut body = json!({
            "model": model,
            "messages": [
                { "role": "system", "content": system },
                { "role": "user", "content": user },
            ],
            "temperature": 0.2,
            "stream": false,
        });

        if let Some(schema) = schema {
            body["response_format"] = json!({
                "type": "json_schema",
                "json_schema": { "name": "out", "strict": true, "schema": schema },
            });
        }

        let res: ChatResponse = self.post("/chat/completions", body)?.json().context("malformed chat response")?;
        res.choices
            .into_iter()
            .next()
            .map(|c| c.message.content)
            .context("no choices returned")
    }
}

File: internal/provider/provider.go
// Package provider is one client for every model runtime.
//
// Ollama, LM Studio, Jan and Msty all speak OpenAI-compatible /v1, so the only
// thing that differs is the port and which models happen to be loaded. Cloud
// BYOK is the same surface with a base URL and a key.
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// LocalEndpoint is a runtime we know how to find without configuration.
type LocalEndpoint struct {
	Name string
	URL  string
}

// Probed in this order; first match wins when nothing is configured.
var LocalEndpoints = []LocalEndpoint{
	{"Ollama", "http://localhost:11434/v1"},
	{"LM Studio", "http://localhost:1234/v1"},
	{"Jan", "http://localhost:1337/v1"},
	{"Msty", "http://localhost:10000/v1"},
}

type Provider struct {
	Name    string
	BaseURL string
	APIKey  string
	http    *http.Client
}

func New(name, baseURL, apiKey string) *Provider {
	return &Provider{
		Name:    name,
		BaseURL: strings.TrimRight(baseURL, "/"),
		APIKey:  apiKey,
		// Local generation on a 24B model genuinely takes a while.
		http: &http.Client{Timeout: 300 * time.Second},
	}
}

type Discovered struct {
	Provider *Provider
	Models   []string
}

// Discover probes the well-known ports and reports what is actually running.
// This is what lets the app say "found LM Studio running Qwen3" on first launch
// instead of presenting an empty endpoint configuration box.
func Discover() []Discovered {
	probe := &http.Client{Timeout: 400 * time.Millisecond}
	var out []Discovered

	for _, ep := range LocalEndpoints {
		res, err := probe.Get(ep.URL + "/models")
		if err != nil {
			continue
		}
		var list struct {
			Data []struct {
				ID string `json:"id"`
			} `json:"data"`
		}
		err = json.NewDecoder(res.Body).Decode(&list)
		res.Body.Close()
		if err != nil {
			continue
		}
		models := make([]string, 0, len(list.Data))
		for _, m := range list.Data {
			models = append(models, m.ID)
		}
		out = append(out, Discovered{New(ep.Name, ep.URL, ""), models})
	}
	return out
}

func (p *Provider) post(path string, body any, into any) error {
	buf, err := json.Marshal(body)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", p.BaseURL+path, bytes.NewReader(buf))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	if p.APIKey != "" {
		req.Header.Set("Authorization", "Bearer "+p.APIKey)
	}

	res, err := p.http.Do(req)
	if err != nil {
		return fmt.Errorf("%s unreachable at %s: %w", p.Name, p.BaseURL, err)
	}
	defer res.Body.Close()

	if res.StatusCode < 200 || res.StatusCode >= 300 {
		var detail bytes.Buffer
		detail.ReadFrom(res.Body)
		return fmt.Errorf("%s returned %s: %s", p.Name, res.Status, strings.TrimSpace(detail.String()))
	}
	return json.NewDecoder(res.Body).Decode(into)
}

func (p *Provider) Embed(model string, inputs []string) ([][]float32, error) {
	if len(inputs) == 0 {
		return nil, nil
	}

	var res struct {
		Data []struct {
			Embedding []float32 `json:"embedding"`
		} `json:"data"`
	}
	if err := p.post("/embeddings", map[string]any{"model": model, "input": inputs}, &res); err != nil {
		return nil, err
	}
	if len(res.Data) != len(inputs) {
		return nil, fmt.Errorf("expected %d embeddings, got %d", len(inputs), len(res.Data))
	}

	out := make([][]float32, len(res.Data))
	for i, d := range res.Data {
		out[i] = d.Embedding
	}
	return out, nil
}

// Chat runs a completion. A non-nil schema enables constrained decoding: small
// local models are unreliable at tool-calling but near-perfect when JSON is
// enforced at the sampler, which is why every extraction path goes through here
// rather than through tool definitions.
func (p *Provider) Chat(model, system, user string, schema map[string]any) (string, error) {
	body := map[string]any{
		"model": model,
		"messages": []map[string]string{
			{"role": "system", "content": system},
			{"role": "user", "content": user},
		},
		"temperature": 0.2,
		"stream":      false,
	}
	if schema != nil {
		body["response_format"] = map[string]any{
			"type": "json_schema",
			"json_schema": map[string]any{
				"name":   "out",
				"strict": true,
				"schema": schema,
			},
		}
	}

	var res struct {
		Choices []struct {
			Message struct {
				Content string `json:"content"`
			} `json:"message"`
		} `json:"choices"`
	}
	if err := p.post("/chat/completions", body, &res); err != nil {
		return "", err
	}
	if len(res.Choices) == 0 {
		return "", fmt.Errorf("no choices returned")
	}
	return res.Choices[0].Message.Content, nil
}
